engine/access/rpc/backend/connection_factory.go

```diff
@@ -61,16 +68,33 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
 	return conn, nil
 }
 
+func (cf *ConnectionFactoryImpl) retrieveConnection(address string) (*grpc.ClientConn, error) {
+	var conn *grpc.ClientConn
+	if res, ok := cf.ConnectionsCache.Get(address); ok {
+		conn = res.(*grpc.ClientConn)
+	}
+	if conn == nil || conn.GetState() != connectivity.Ready {
+		conn, err := cf.createConnection(address, cf.ExecutionNodeGRPCTimeout)
+		if err != nil {
+			return nil, err
+		}
+		cf.ConnectionsCache.Add(address, conn)
+	}
```

Comment:
I think it would be helpful to add metrics that track connection reuse. Maybe metrics for total connections and connections reused.